flink-streaming-java/src/test/java/org/apache/flink/streaming/util/WindowingTestHarness.java

```diff
@@ -0,0 +1,217 @@
+								Trigger<? super IN, ? super W> trigger,
+								long allowedLateness) {
+
+		ListStateDescriptor<IN> windowStateDesc = new ListStateDescriptor<>("window-contents",
+			inputType.createSerializer(new ExecutionConfig()));
+
+		WindowOperator<K, IN, Iterable<IN>, IN, W> operator =
+			new WindowOperator<>(
+				windowAssigner,
+				windowSerializer,
+				keySelector,
+				keyType.createSerializer(executionConfig),
+				windowStateDesc,
+				new InternalIterableWindowFunction<>(new PassThroughFunction()),
+				trigger,
+				allowedLateness);
+
+		operator.setInputType(inputType, executionConfig);
+
+		timeServiceProvider = new TestTimeServiceProvider();
+		testHarness = new OneInputStreamOperatorTestHarness<>(operator, executionConfig, timeServiceProvider);
+		testHarness.configureForKeyedStream(keySelector, keyType);
+	}
+
+	/**
+	 * Simulates the processing of a new incoming element.
+	 */
+	public void processElement(IN element, long timestamp) throws Exception {
+		openOperator();
+		testHarness.processElement(new StreamRecord<>(element, timestamp));
+	}
+
+	/**
+	 * Simulates the processing of a new incoming watermark.
+	 */
+	public void processWatermark(long timestamp) throws Exception {
+		openOperator();
+		testHarness.processWatermark(new Watermark(timestamp));
+	}
+
+	/**
+	 * Sets the current processing time to {@code timestamp}.
+	 * This is useful when working on processing time.
+	 */
+	public void setProcessingTime(long timestamp) throws Exception {
+		openOperator();
+		timeServiceProvider.setCurrentTime(timestamp);
+	}
+
+	/**
+	 * Gets the current output of the windowing operator, as produced by the
+	 * synergies between the window assigner and the trigger. This will also
+	 * contain the received watermarks.
+	 */
+	public ConcurrentLinkedQueue<Object> getOutput() throws Exception {
+		return testHarness.getOutput();
+	}
+
+	/**
+	 * Closes the testing window operator.
+	 */
+	public void close() throws Exception {
+		if (isOpen) {
+			testHarness.close();
+			isOpen = false;
+		}
+	}
+
+	/**
+	 * Adds a watermark to the expected output.
+	 * The expected output should contain the elements and watermarks that we expect the output of the operator to
+	 * contain, in the correct order. This will be used to check if the produced output is the expected one, and
+	 * thus determine the success or failure of the test.
+	 */
+	public void addExpectedWatermark(long timestamp) {
+		expectedOutputs.add(new Watermark(timestamp));
+	}
+
+	/**
+	 * Adds an element to the expected output.
+	 * The expected output should contain the elements and watermarks that we expect the output of the operator to
+	 * contain, in the correct order. This will be used to check if the produced output is the expected one, and
+	 * thus determine the success or failure of the test.
+	 */
+	public void addExpectedElement(IN element, long timestamp) {
+		expectedOutputs.add(new StreamRecord<>(element, timestamp));
+	}
+
+	/**
```

Comment:
We should maybe mention that this internally uses an assert and the test thus fails when the expected output does not match the actual output.